rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,477 @@
+            nonce.as_ptr(),
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data_ptr,
+            auth_data_len,
+            cipher_text.as_ptr(),
+            plain_text.as_mut_ptr(),
+            cipher_text.len() as symcrypt_sys::SIZE_T,
+            tag.as_ptr(),
+            tag.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(plain_text),
+            err => Err(err.into()),
+        }
+    }
+}
+
+// SymCrypt requires pointer address to stay static through the scope, passing expand_key as an out
+// parameter to maintain static pointer address
+fn gcm_expand_key(
+    key: &[u8],
+    expanded_key: &mut symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+    cipher: *const symcrypt_sys::SYMCRYPT_BLOCKCIPHER,
+) -> Result<(), SymCryptError> {
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptGcmExpandKey(
+            expanded_key,
+            cipher,
+            key.as_ptr(),
+            key.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(()),
+            err => Err(err.into()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::block_ciphers::BlockCipherType;
+    #[test]
+    fn test_stateless_gcm_encrypt_no_ad() {
+        let p_key = hex::decode("feffe9928665731c6d6a8f9467308308").unwrap();
+        let nonce = hex::decode("cafebabefacedbaddecaf888").unwrap();
+        let expected_tag = "4d5c2af327cd64a62cf35abd2ba6fab4";
+        let expected_result = "42831ec2217774244b7221b784d0d49ce3aa212f2c02a4e035c17e2329aca12e21d514b25466931c7d8f6a5aac84aa051ba30b396a0aac973d58e091473f5985";
+        let pt = hex::decode("d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a721c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b391aafd255").unwrap();
+
+        let nonce_array: [u8; 12] = nonce.as_slice().try_into().unwrap();
+
+        let cipher = BlockCipherType::AesBlock;
+        let (dst, tag) = gcm_encrypt(&p_key, &nonce_array, None, &pt, 16, cipher).unwrap();
+
+        assert_eq!(expected_result, hex::encode(dst));
+        assert_eq!(expected_tag, hex::encode(tag));
+    }
+
+    #[test]
+    fn test_stateless_gcm_encrypt_ad() {
+        let p_key = hex::decode("feffe9928665731c6d6a8f9467308308").unwrap();
+        let nonce = hex::decode("cafebabefacedbaddecaf888").unwrap();
+        let auth_data = hex::decode("feedfacedeadbeeffeedfacedeadbeefabaddad2").unwrap();
+        let expected_tag = "5bc94fbc3221a5db94fae95ae7121a47";
+        let expected_result = "42831ec2217774244b7221b784d0d49ce3aa212f2c02a4e035c17e2329aca12e21d514b25466931c7d8f6a5aac84aa051ba30b396a0aac973d58e091";
+        let pt = hex::decode("d9313225f88406e5a55909c5aff5269a86a7a9531534f7da2e4c303d8a318a721c3c0c95956809532fcf0e2449a6b525b16aedf5aa0de657ba637b39").unwrap();
+
+        let nonce_array: [u8; 12] = nonce.as_slice().try_into().unwrap();
```

Comment:
>let nonce_array: [u8; 12] = nonce.as_slice().try_into().unwrap(); [](http://example.com/codeflow?start=8&length=65)

Why is this necessary?  Can't you do the following:

&nonce[0 .. 12]

Reply:
`&nonce[0 .. 12]` produces `&[u8]`, we need ` [u8; 12] `

Reply:
hex::decode creates the Vec